lib/usb/usb_process.js

```diff
@@ -1,236 +1,271 @@
-  this.stdout.resume();
-  this.stderr.resume();
-
-  this.stdout.signal = function(signalToSend) {
-    switch(signalToSend) {
-      case 'KILL':
-        self.kill(9);
-        break;
-    }
-  }
-
-  // When we receive a stderr write message, we'll want to push it
-  self.on('write', function(packet) {
-    // Push the suggested data into the suggested stream
-    packet.stream.push(packet.data);
-
-    var ackFunc = (packet.stream == self.stdout ? protocol.stdoutAck : protocol.stderrAck);
-
-    self._readableAckHelper(packet.stream , ackFunc, packet.data.length);
-  });
+  // 4 remote process streams
+  this.control = new RemoteWritableStream(id, daemon, protocol.controlWrite, protocol.controlClose);
+  this.stdin = new RemoteWritableStream(id, daemon, protocol.stdinWrite, protocol.stdinClose);
+  this.stdout = new RemoteReadableStream(id, daemon, protocol.stdoutAck);
+  this.stderr = new RemoteReadableStream(id, daemon, protocol.stderrAck);
 
   // Once the process is killed
   self.once('death', function(data) {
-    self.active = false;
-    // Set the exit code
-    self.exitCode = data.arg;
-    // Close the input streams
+    // Close the input streams because the remote proc can't handle data
     self.control.end();
     self.stdin.end();
-    // Clean up remote resources
-    self.close();
+
+    // Mark this process as inactive
+    self.active = false;
   });
 
-  this.close = function() {
+  // Tell the USBDaemon to close the remote process
+  self.close = function() {
     // Write the close command
     daemon.write(protocol.closeProcess(self.id));
-    self.once('close', function() {
-
-      // End the output streams
-      self.stdout.push(null);
-      self.stderr.push(null);
-    })
   };
 
-  this.kill = function(signal) {
+  // Send a signal to the remote process
+  self.kill = function(signal) {
+    // Write the signal command
     daemon.write(protocol.killProcess(self.id, signal));
   }
 
-
-  this.control.on('drain', function() {
-    self._drainBackPressure(self.control, protocol.controlWrite);
-  });
-
-
-  this.stdin.on('drain', function() {
-    self._drainBackPressure(self.stdin, protocol.stdinWrite);
-  });
-
-  this.control._write = function(chunk, enc, callback) {
-    // Add this chunk to the backpressure array
-    self.control.backPressure.push({buffer: chunk, callback: function controlSent() {
-      // Tell the daemon that we have finished writing to the control stream
-      daemon.write(protocol.controlClose(self.id));
-      // Call the callback
-      callback && callback();
-    }});
-
-    if (self.control.credit) {
-      self._drainBackPressure(self.control, protocol.controlWrite);
+  // Checks if this process is ready to be closed
+  // and closes it if it can
+  self.closeProcessIfReady = function() {
+    // If this process hasn't been closed before
+    // and all of the streams are closed
+    if (self.active && self.stdout.closed
```

Comment:
Should this be `!self.active`? (i.e. the process has exited).

Also, I'm not sure about including stdin here, since closing stdin is controlled by the CLI side. It's fine if the user of this library is careful to close stdin when done writing. Programs that accept input on stdin like `tar` will themselves not exit until their stdin is closed, so those are fine. But when you run a program that doesn't accept input, you may not touch stdin at all and forget to close it, leaking the remote process.